File: domain/src/main/java/org/openmedstack/eventstore/IDetectConflicts.java
package org.openmedstack.eventstore;

import org.openmedstack.Tuple;

import java.util.Arrays;
import java.util.HashMap;
import java.util.function.Function;
import java.util.stream.Collectors;

public interface IDetectConflicts
{
    <TUncommitted, TCommitted> void register(Function<Tuple<TUncommitted, TCommitted>, Boolean> handler);

    Boolean conflictsWith(Object[] uncommittedEvents, Object[] committedEvents);
}

File: domain/src/main/kotlin/org/openmedstack/eventstore/IDetectConflicts.kt
package org.openmedstack.eventstore

import org.openmedstack.Tuple
import java.util.function.Function

interface IDetectConflicts {
    fun <TUncommitted, TCommitted> register(handler: Function<Tuple<TUncommitted, TCommitted>, Boolean>)
    fun conflictsWith(uncommittedEvents: Array<Any>, committedEvents: Array<Any>): Boolean
}